samples/spring-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/ShoppingCartService.java

```diff
@@ -0,0 +1,116 @@
+package com.example.shoppingcart;
+
+import com.example.shoppingcart.domain.ShoppingCart;
+import com.example.shoppingcart.domain.ShoppingCart.LineItem;
+import com.example.shoppingcart.domain.ShoppingCartEvent;
+import kalix.javasdk.eventsourcedentity.EventSourcedEntity;
+import kalix.javasdk.eventsourcedentity.EventSourcedEntityContext;
+import kalix.springsdk.annotations.Entity;
+import kalix.springsdk.annotations.EventHandler;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+// tag::class[]
+@Entity(entityKey = "cartId", entityType = "shopping-cart")
+@RequestMapping("/cart/{cartId}")
+public class ShoppingCartService extends EventSourcedEntity<ShoppingCart> {
+
+  private final String entityId;
+
+  public ShoppingCartService(EventSourcedEntityContext context) { this.entityId = context.entityId(); }
+
+  @Override
+  public ShoppingCart emptyState() { // <2>
+    return new ShoppingCart(entityId, Collections.emptyList());
+  }
+  // end::class[]
+
+  // tag::addItem[]
+  @PostMapping("/add")
+  public Effect<String> addItem(@RequestBody LineItem item) {
+    if (item.quantity() <= 0) { // <1>
+      return effects().error("Quantity for item " + item.productId() + " must be greater than zero.");
+    }
+
+    var event = new ShoppingCartEvent.ItemAdded(item); // <2>
+
+    return effects()
+        .emitEvent(event) // <3>
+        .thenReply(newState -> "OK"); // <4>
+  }
+  // end::addItem[]
+
+  @PostMapping("/items/{productId}/remove")
+  public Effect<String> removeItem(@PathVariable String productId) {
+    if (findItemByProductId(currentState(), productId).isEmpty()) {
+      return effects()
+          .error(
+              "Cannot remove item " + productId + " because it is not in the cart.");
+    }
+
+    var event = new ShoppingCartEvent.ItemRemoved(productId);
+
+    return effects()
+        .emitEvent(event)
+        .thenReply(newState -> "OK");
+  }
+
+  // tag::getCart[]
+  @GetMapping
+  public Effect<ShoppingCart> getCart() {
+    return effects().reply(currentState());
+  }
+  // end::getCart[]
+
+  // tag::itemAdded[]
+  @EventHandler
+  public ShoppingCart itemAdded(ShoppingCartEvent.ItemAdded itemAdded) {
+    var item = itemAdded.item();
+    var lineItem = updateItem(item, currentState());
+    List<LineItem> lineItems =
+        removeItemByProductId(currentState(), item.productId());
+    lineItems.add(lineItem);
+    lineItems.sort(Comparator.comparing(LineItem::productId));
+    return currentState().withItems(lineItems);
```

Comment:
To emphasize that the `ShoppingCart` is the model and that `ShoppingCartService` is just a service around it, we can move this kind of logic to `ShoppingCart`. 

This is not possible in the gRPC SDKs, because the models are generated and we can't add logic to them. So users have to create anemic models with logic in the services. 

So the pattern that I had in mind was to move as much logic as possible to the model and have the service as a thin wrapper/facade between the model and kalix.

Reply:
Yeah, thought the same earlier today when looking at what you've done for the customer registry sample... I think it's a better pattern. Will update. Thanks

Reply:
Yep, I think is a good way forward. 

One thing that still not perfect, is that the Service extends something called `EventSourcedEntity` and that's kind of weird.

Also, when we subscribe to it from a view, we end up with:

`@Subscribe.EventSourcedEntity(ShoppingCartService.class)` 

That's also weird.